### PR TITLE
Add Feature #49 - Pull Request

### PR DESCRIPTION
**Pull request by amondel2**
This will allow a person to enable or disabling sending emails when forgetting a password or registering.

**Parameter Documentation:**
1. validationUserLookUpProperty - This is the name of the property in Domain Object for finding a user. It is used by the forgotPasswordExtraValidation for ensuring the items being checked belong to the user whos password that is trying to be reset.
2. requireForgotPassEmailValidation – This property sets if an Email should go out for resetting the password
3. requireEmailValidation – This property sets if an Email should go out when a new user signs up. If this is false a user should add some additional step.
4. forgotPasswordExtraValidation – This is an List which contains 4 other parameters
1. domain – This is Domain object that contains the question ( if labelDomain is present ) as will as the answer
2. labelMessage - If labelDomain is not present then this can be used as a message property to ask a question (ie if your questions are static).
3. prop – This is used a the property of the domain object to determine if the answer is correct.

There are three new default security and one additional new property not put in the default config to enable the security questions, which is separated from the if the e-mail is sent or not. This means you can have both security questions and email enabled, neither, or just one or the other.
```

security {
	ui {
		forgotPassword {
			validationUserLookUpProperty = 'user'    
			requireForgotPassEmailValidation = 'true'
		}
		register {
			requireEmailValidation = 'true' //set this to false if you don't want an e-mail validation
		}
	}
}
```

In you projects application.yml

```

grails:
    plugin:
        springsecurity:
            ui:
                forgotPassword:
                    forgotPasswordExraValidation:
                             -
                                labelDomain: myQuestion
                                 prop: myAnswer
                             -
                                   labelDomain: myQuestion2
                                 labelMessage: securityvalidations.labelMessage.label1
                                 prop: myAnswer2
```

This PR is for Issue https://github.com/grails-plugins/grails-spring-security-ui/issues/49

### CODE REVIEW

File: examples/extended/grails-app/conf/application.yml [hunk 1]
```diff
@@ -193,4 +193,17 @@ grails:
             userLookup:
                 authorityJoinClassName: test.UserRole
                 userDomainClassName: test.User
-            useSwitchUserFilter: true
\ No newline at end of file
+            useSwitchUserFilter: true
+            ui:
```
Comment (ddelponte) [inline examples/extended/grails-app/conf/application.yml]: Shouldn't this be organized under `springsecurity`?

File: examples/extended/grails-app/i18n/messages.properties [hunk 2]
```diff
@@ -54,3 +54,4 @@ typeMismatch.java.lang.Short=Property {0} must be a valid number
 typeMismatch.java.math.BigDecimal=Property {0} must be a valid number
 typeMismatch.java.math.BigInteger=Property {0} must be a valid number
 typeMismatch=Property {0} is type-mismatched
+securityvalidations.labelMessage.label1=Count to six minus 1
```
Comment (ddelponte) [inline examples/extended/grails-app/i18n/messages.properties]: You may want to add some physical space (new line) between this property and the data binding error properties above to help communicate that they are not related.

File: plugin/src/main/groovy/grails/plugin/springsecurity/ui/strategy/DefaultRegistrationCodeStrategy.groovy [hunk 3]
```diff
@@ -40,6 +41,14 @@ class DefaultRegistrationCodeStrategy implements RegistrationCodeStrategy {
 		springSecurityUiService.register user, password, salt
 	}
 
+	def verifyRegistration(String token) {
```
Comment (ddelponte) [inline plugin/src/main/groovy/grails/plugin/springsecurity/ui/strategy/DefaultRegistrationCodeStrategy.groovy]: Why was verification moved out of the controller and into the default strategy?

File: plugin/grails-app/controllers/grails/plugin/springsecurity/ui/RegisterController.groovy [hunk 4]
```diff
@@ -118,79 +137,116 @@ class RegisterController extends AbstractS2UiController implements GrailsConfigu
 	}
 
 	def verifyRegistration() {
+		redirectVerifyRegistration(uiRegistrationCodeStrategy.verifyRegistration(params.t))
```
Comment (ddelponte) [inline plugin/grails-app/controllers/grails/plugin/springsecurity/ui/RegisterController.groovy]: Why was verification moved out of the controller and into the default strategy?

File: examples/extended/grails-app/views/register/secuirtyQuestions.gsp [hunk 5]
```diff
@@ -0,0 +1,19 @@
+<html>
```
Comment (ddelponte) [inline examples/extended/grails-app/views/register/secuirtyQuestions.gsp]: The name of this view is mispelled.

Instead of `secuirtyQuestions` it should be `securityQuestions`.

This should also be corrected anywhere it is referenced.

File: plugin/src/main/groovy/grails/plugin/springsecurity/ui/strategy/DefaultRegistrationCodeStrategy.groovy [hunk 6]
```diff
@@ -40,6 +41,14 @@ class DefaultRegistrationCodeStrategy implements RegistrationCodeStrategy {
 		springSecurityUiService.register user, password, salt
```
Comment (ddelponte) [inline plugin/src/main/groovy/grails/plugin/springsecurity/ui/strategy/DefaultRegistrationCodeStrategy.groovy]: I've tested this locally and, by default, if `requireForgotPassEmailValidation = false`, newly registered users are created with a locked account.

Was this intentional?

File: examples/extended/grails-app/conf/application.yml (hunk 1 above)
Comment (amondel2) [reply examples/extended/grails-app/conf/application.yml]: No, I don't think so based on where the other items are configured in DefaultUiSecurityConfig.groovy in the main plugin.

File: examples/extended/grails-app/i18n/messages.properties (hunk 2 above)
Comment (amondel2) [reply examples/extended/grails-app/i18n/messages.properties]: Agree will fix!

File: plugin/grails-app/controllers/grails/plugin/springsecurity/ui/RegisterController.groovy (hunk 4 above)
Comment (amondel2) [reply plugin/grails-app/controllers/grails/plugin/springsecurity/ui/RegisterController.groovy]: It also called on line 88 of this controller when Email Validation is not required.  Previously the link in the email would have been the only way to trigger this action.

File: plugin/grails-app/controllers/grails/plugin/springsecurity/ui/RegisterController.groovy [hunk 7]
```diff
@@ -79,9 +79,28 @@ class RegisterController extends AbstractS2UiController implements GrailsConfigu
...
+			redirectVerifyRegistration(uiRegistrationCodeStrategy.verifyRegistration(registrationCode.token))
+		}
+
+
+
+	}
+
+	protected void redirectVerifyRegistration(def rtn) {
+		if (rtn?.flashmsg) {
+			if(rtn?.flashType == 'error')
```
[the remark this answers is not in the document]
Comment (amondel2) [reply plugin/grails-app/controllers/grails/plugin/springsecurity/ui/RegisterController.groovy]: Fixing

File: examples/extended/grails-app/conf/application.yml [hunk 8]
```diff
@@ -193,4 +193,17 @@ grails:
...
+                forgotPassword:
+                    forgotPasswordExraValidation:
```
[the remark this answers is not in the document]
Comment (amondel2) [reply examples/extended/grails-app/conf/application.yml]: Yes and in So many places..will fix

File: plugin/src/main/groovy/grails/plugin/springsecurity/ui/strategy/DefaultRegistrationCodeStrategy.groovy (hunk 3 above)
Comment (amondel2) [reply plugin/src/main/groovy/grails/plugin/springsecurity/ui/strategy/DefaultRegistrationCodeStrategy.groovy]: See comment below.

File: examples/extended/grails-app/views/register/secuirtyQuestions.gsp (hunk 5 above)
Comment (amondel2) [reply examples/extended/grails-app/views/register/secuirtyQuestions.gsp]: Will Fix

File: plugin/src/main/groovy/grails/plugin/springsecurity/ui/strategy/DefaultRegistrationCodeStrategy.groovy (hunk 6 above)
Comment (amondel2) [reply plugin/src/main/groovy/grails/plugin/springsecurity/ui/strategy/DefaultRegistrationCodeStrategy.groovy]: I will do some testing later one but it seems that this parameter should not effect if the newly created accounts are set to be locked out.  The configuration of requireEmailValidation should have some effect on this because it looks like when new accounts are created they are supposed to be locked out until the e-mail is clicked.  If this was set to false then I would think it would be unlocked sine the token step happens behind the scene.

**Comment by amondel2** [comment]
I think I address all the code review items..I am still reviewing the documentation but I did add a good bit in the configuration sections.

**Comment by amondel2** [comment]
@ddelponte you may want to start testing. I should finish the documentation in the next two weeks...it's a matter of getting time not that it needs much..maybe 2-4 hours.

File: plugin/src/main/groovy/grails/plugin/springsecurity/ui/strategy/DefaultRegistrationCodeStrategy.groovy (hunk 6 above)
Comment (ddelponte) [reply plugin/src/main/groovy/grails/plugin/springsecurity/ui/strategy/DefaultRegistrationCodeStrategy.groovy]: I've tested this again and there appears to be a bug.

Yes, the token verification step does set `User.accountLocked = false`, but at some point after that it is persisted to the database as `User.accountLocked = true`

To duplicate this bug, create a simple grails app which utilizes your plugin changes, set `requireForgotPassEmailValidation = false` and register a new user.

Once registered, login as an admin and user the `UserController` to search for the new user and view their details.  You will see that their account is locked.

Comment (ddelponte) [reply plugin/src/main/groovy/grails/plugin/springsecurity/ui/strategy/DefaultRegistrationCodeStrategy.groovy]: @amondel2 I found the cause of the issue.  

The `User.accountLocked` field will be true if the default role name being associated with the user doesn't exist.

This is not a bug, but the code should be improved to avoid this.

I'll writeup a separate issue to tackle this improvement since it is unrelated to your changes.

Thanks!

Comment (amondel2) [reply plugin/src/main/groovy/grails/plugin/springsecurity/ui/strategy/DefaultRegistrationCodeStrategy.groovy]: Thank you for doing the RCA!

File: plugin/grails-app/controllers/grails/plugin/springsecurity/ui/RegisterController.groovy [hunk 9]
```diff
@@ -118,79 +136,116 @@ class RegisterController extends AbstractS2UiController implements GrailsConfigu
...
 		}
 
-		def user = findUserByUsername(forgotPasswordCommand.username)
-		if (!user) {
-			forgotPasswordCommand.errors.rejectValue 'username', 'spring.security.ui.forgotPassword.user.notFound'
-			return [forgotPasswordCommand: forgotPasswordCommand]
-		}
 
+	protected def processForgotPasswordEmail(forgotPasswordCommand,user){
 		String email = uiPropertiesStrategy.getProperty(user, 'email')
```
Comment (ddelponte) [inline plugin/grails-app/controllers/grails/plugin/springsecurity/ui/RegisterController.groovy]: Line #216: `String email = uiPropertiesStrategy.getProperty(user, 'email')` should be moved inside of the `if` statement at line 217, `if(requireForgotPassEmailValidation) {` to avoid the error `ERROR --- [io-8080-exec-10] g.p.s.ui.SpringSecurityUiService         : Attempted to read non-existent property email in User(username:me)` if email is not being used.

It should look something like:

```
	protected def processForgotPasswordEmail(forgotPasswordCommand,user){
		if(requireForgotPassEmailValidation) {
		        String email = uiPropertiesStrategy.getProperty(user, 'email')
			if (!email) {
.
.
.

```

**Comment by ddelponte** [comment]
I've completed testing and it appears to be working great!  

All that's left to do is the complete the documentation 👍 ✨ 

Once the docs are done, I'll review and get the code merged to master.

Thanks again for this contribution.  Your hard work is much appreciated.

**Comment by ddelponte** [comment]
**Additional Tasks**

I thought of two more things that need to be completed before this is ready for public consumption:

1. The answers to security questions should be hashed before persisting to the database.
2. A new script should be created which 
   - generates the security question domain class
   - registers the required listener to encode the answers to security questions
   - enables security question functionality by modifying `application.yml`

**New Security Question Listener**

The security question listener may be implemented similarly to what's being done for password encoding:

1. create a listener for the domain class associated with forgot password questions (Profile in this example)
2. encode all answers before updating/persisting the entity
3. modify the code that verifies answers to compare hashes

More info is available at http://gorm.grails.org/latest/hibernate/manual/index.html#advancedGORMFeatures under the Custom Event Listeners heading

**New s2-create-security-questions` script**

create a new script
   - maybe something like `s2-create-security-questions <classname> <# of questions>` that generates the domain class with the appropriate numbers of question/answer fields
   - For example, `s2-create-security-questions com.mycompany.Profile 4` would generate

 ```
package com.mycompany.myapp

class Profile implements Serializable {

    private static final long serialVersionUID = 12

    String myQuestion1
    String myAnswer1
    String myQuestion2
    String myAnswer2
    String myQuestion3
    String myAnswer3
    String myQuestion4
    String myAnswer4
    User user
    static belongsTo = [user: User]
    static constraints = {
    }
}   
   ```
1. the script should modify `grails-app/conf/resources.groovy` by registering the new listener
1. the script should add the settings required to enable the security question feature to `application.yml`

File: plugin/grails-app/controllers/grails/plugin/springsecurity/ui/RegisterController.groovy (hunk 9 above)
Comment (amondel2) [reply plugin/grails-app/controllers/grails/plugin/springsecurity/ui/RegisterController.groovy]: This will be fixed in my next pull request

**Comment by amondel2** [comment]
@ddelponte  In working on the Additional task do you think I over complicated this by adding the ability to have the each security questions/answers in multiple domain objects?  I noticed the complexity it will add when looking at these scripts and encoders it would be much easier to pull this out into an non arrayed configuration and only have this in one domain class and the change to do that is fairly minor?

**Comment by amondel2** [comment]
@ddelponte quick status update i have the listener working with unit test.  I did go to the one domain object property to make this easier.  I also used a listener by using what was demonstrated at the following link http://guides.grails.org/gorm-event-listeners/guide/index.html.  I have to fix one case I fubared when doing a merge, but i am planning to get to that to tonight/tomorrow and then I will work on the script.

**Comment by ddelponte** [comment]
@amondel2 apologies for the late response.

I agree with the one domain object property approach.  Definitely less complicated and more manageable.  

And again, great work on this new feature.  Your hard work is appreciated.

**Comment by amondel2** [comment]
@ddelponte quick status update.  I though I was done the scripting file but during my review in preparation for the documentation steps on how to use the script I realized I needed two additional items, 1 the user class in my template had been hard coded, so an extra parameter to tell me what they are using, this is done in my branch.   The second is that that Security Questions Controllers and view were all dependent on scaffolding and I don't want to create a new dependency.   I am now working on creating the views and controller in a way that a) makes them look similar to other UI screens and b) doesn't create additional dependencies.

**Comment by amondel2** [comment]
@ddelponte Hopefully this is my last status update for this feature. 

The code is finished.  I didn't implement search..I think that can be an additional feature added later, right now it just goes to the list/index page.   Additionally, in the future we may want to think about adding this as a new tab in the user section.  I think this out of scope for this as well, but would work on this as a separate feature after this is completed, 

The test all passing https://travis-ci.org/amondel2/grails-spring-security-ui/builds/403972058

Documentation is the last task that needs to be completed before I merge up again.

**Comment by ddelponte** [comment]
Thank-you @amondel2 .  Great work!

I will review once you have committed the final documentation changes. ✨ 

It also looks like your latest commit broke the build:
https://travis-ci.org/grails-plugins/grails-spring-security-ui/builds/398878981?utm_source=github_status&utm_medium=notification

**Comment by amondel2** [comment]
@ddelponte I am ready for a new code review/testing which would include the documentation review.  Let me know what I missed from the code review of myself :) 
Also I am no longer the breaker of build :trophy:  :smile:

File: examples/extended/grails-app/i18n/messages.properties [hunk 10]
```diff
@@ -54,3 +54,7 @@ typeMismatch.java.lang.Short=Property {0} must be a valid number
 typeMismatch.java.math.BigDecimal=Property {0} must be a valid number
 typeMismatch.java.math.BigInteger=Property {0} must be a valid number
 typeMismatch=Property {0} is type-mismatched
+
+#Example of how to use the label for extra validation
```
Comment (ddelponte) [inline examples/extended/grails-app/i18n/messages.properties]: Would it be better if the comment read as follows?:

```
# Example of how to use the labelMessage for extra validation
```

File: examples/extended/grails-app/services/test/TestDataService.groovy [hunk 11]
```diff
@@ -158,7 +165,7 @@ class TestDataService {
 	}
 
 	protected save(instance) {
-		instance.save(failOnError: true)
+		instance.save(failOnError: true, flush: true)
```
Comment (ddelponte) [inline examples/extended/grails-app/services/test/TestDataService.groovy]: Why does this need `flush: true`?  It's best to avoid flushing if at all possible.

File: examples/extended/src/integration-test/groovy/page/AbstractSecurityPage.groovy [hunk 12]
```diff
@@ -9,10 +9,11 @@ abstract class AbstractSecurityPage extends Page {
 	}
 
 	protected boolean assertContentContains(String expected) {
-		assert $().text().contains(expected)
+		assert $().text().contains(expected) || $().text().contains(expected)
```
Comment (ddelponte) [inline examples/extended/src/integration-test/groovy/page/AbstractSecurityPage.groovy]: Duplicate check is being performed.

```
assert $().text().contains(expected) || $().text().contains(expected)
```

should be

```
assert $().text().contains(expected)
```

File: examples/extended/src/integration-test/groovy/spec/RegisterSpec.groovy [hunk 13]
```diff
@@ -86,38 +93,51 @@ class RegisterSpec extends AbstractSecuritySpec {
...
+		when:
+		go 'register/resetPassword?t=123'
+
+		then:
+		assertHtmlContains 'Sorry, we have no record of that request, or it has expired'
 
 		when:
 		to RegisterPage
+
+        then:
```
Comment (ddelponte) [inline examples/extended/src/integration-test/groovy/spec/RegisterSpec.groovy]: indentation and formatting is a bit off.  Would you be able to make it more consistent?

File: examples/extended/src/integration-test/groovy/test/ProfileServiceSpec.groovy [hunk 14]
```diff
@@ -0,0 +1,73 @@
...
+
+@Integration
+@Rollback
+class ProfileServiceSpec extends Specification {
+
+    ProfileService profileService
+    SessionFactory sessionFactory
+
+    private Long setupData() {
+        // TODO: Populate valid domain instances and return a valid ID
```
Comment (ddelponte) [inline examples/extended/src/integration-test/groovy/test/ProfileServiceSpec.groovy]: Can all of this commented out code be deleted?

**Comment by ddelponte** [comment]
@amondel2 , I believe a found a bug in the `s2ui-create-challenge-questions` script.

**Steps to reproduce**
1. I have  pre-existing domain classes in a package:
image

2. I run the script as follows:

```
s2ui-create-challenge-questions com.mycompany.myapp ChallengeQuestion com.mycompany.myapp.User 2
```

**Actual results**
3. the new domain class, `ChallengeQuestion` is not created in the existing package (the same package as User)

image

**Expected results**
I would expect the domain class `ChallengeQuestion` to be created in the existing directory `com > mycompany > myapp` and not in a new directory named `com.mycompany.myapp`

File: examples/extended/src/integration-test/groovy/page/AbstractSecurityPage.groovy (hunk 12 above)
Comment (amondel2) [reply examples/extended/src/integration-test/groovy/page/AbstractSecurityPage.groovy]: fixed in next commit..I thought I got that before I checked it in

File: examples/extended/src/integration-test/groovy/spec/RegisterSpec.groovy (hunk 13 above)
Comment (amondel2) [reply examples/extended/src/integration-test/groovy/spec/RegisterSpec.groovy]: Add tabs instead of spaces

**Comment by ddelponte** [comment]
Another bug in the `s2ui-create-challenge-questions` script is that `Profile`, the domain class you've used in your examples, is hardcoded in the listener service:

image

It's hardcoded in:

1. `onProfilePreInsert`
2. `onProfilePreUpdate`
3. `Profile prof = ((ChallengeQuestion) event.entityObject)`

File: examples/extended/grails-app/services/test/TestDataService.groovy (hunk 11 above)
Comment (amondel2) [reply examples/extended/grails-app/services/test/TestDataService.groovy]: fixed in next commit

File: examples/extended/grails-app/i18n/messages.properties (hunk 10 above)
Comment (amondel2) [reply examples/extended/grails-app/i18n/messages.properties]: fixed in next commit

File: examples/extended/src/integration-test/groovy/test/ProfileServiceSpec.groovy (hunk 14 above)
Comment (amondel2) [reply examples/extended/src/integration-test/groovy/test/ProfileServiceSpec.groovy]: fixed in next commit

File: docs/src/docs/forgotPassword.adoc [hunk 15]
```diff
@@ -38,8 +40,96 @@ The post-reset destination url is configurable in `grails-app/conf/application.g
...
+    springsecurity:
+     ui:
+      forgotPassword:
+          forgotPasswordExtraValidationDomainClassName: com.mycompany.Profile
+----
+
+
+=== Mail configuration
+
+By default the plugin uses the http://grails.org/plugin/mail[Mail] plugin to send emails, but only if it installed. This is configurable by registering your own `MailStrategy` implementation - see [the section on configuration|guide:customization] for more information. The plugin assumes that the Mail plugin and an SMTP server are already configured.
```
[the remark this answers is not in the document]
Comment (amondel2) [reply docs/src/docs/forgotPassword.adoc]: both fixed in next commit..I

File: docs/src/docs/forgotPassword.adoc [hunk 16]
```diff
@@ -38,8 +40,96 @@ The post-reset destination url is configurable in `grails-app/conf/application.g
...
+               labelMessage: securityvalidations.labelMessage.label1
+               prop: myAnswer2
+----
+
+If the above is configured you will see a menu item to List and Create Chanallage questions for each
+user.
+
+
+To use the challenge questions functionality please set the domain class which will hold the answers with the
+'forgotPasswordExtraValidationDomainClassName' property.  This is Domain object that contains any reference that is needed by prop (question) and labelDomain (answer).
```
[the remark this answers is not in the document]
Comment (amondel2) [reply docs/src/docs/forgotPassword.adoc]: fixed in next commit

File: docs/src/docs/forgotPassword.adoc [hunk 17]
```diff
@@ -38,8 +40,96 @@ The post-reset destination url is configurable in `grails-app/conf/application.g
...
+         forgotPasswordExtraValidation:
+            -
+               labelDomain: myQuestion
+               prop: myAnswer
+            -
+               labelMessage: securityvalidations.labelMessage.label1
+               prop: myAnswer2
+----
+
+If the above is configured you will see a menu item to List and Create Chanallage questions for each
```
[the remark this answers is not in the document]
Comment (amondel2) [reply docs/src/docs/forgotPassword.adoc]: It was Challenge to spell this correctly but only once :)
Fixed in next commit

File: docs/src/docs/forgotPassword.adoc [hunk 18]
```diff
@@ -38,8 +40,96 @@ The post-reset destination url is configurable in `grails-app/conf/application.g
...
+|labelMessage
+|If labelDomain is not present then this can be used as a message property to ask a question (ie if your questions are static).
+
+|prop
+|This is used a the property of the domain object to determine if the answer is correct.
+|===
+
+
+This can be customized in either `grails-app/conf/application.groovy` or `grails-app/conf/application.yml` but the example below is
+for the yml file.
```
[the remark this answers is not in the document]
Comment (amondel2) [reply docs/src/docs/forgotPassword.adoc]: fixed

File: docs/src/docs/forgotPassword.adoc [hunk 19]
```diff
@@ -38,8 +40,96 @@ The post-reset destination url is configurable in `grails-app/conf/application.g
...
+|Header 1 |Header 2
+
+|labelDomain
+|This is the property in the domain object that will used to ask the question
+
+|labelMessage
+|If labelDomain is not present then this can be used as a message property to ask a question (ie if your questions are static).
+
+|prop
+|This is used a the property of the domain object to determine if the answer is correct.
```
[the remark this answers is not in the document]
Comment (amondel2) [reply docs/src/docs/forgotPassword.adoc]: fixed in next commit

File: docs/src/docs/forgotPassword.adoc [hunk 20]
```diff
@@ -38,8 +40,96 @@ The post-reset destination url is configurable in `grails-app/conf/application.g
...
+To use this option it is recommended to use the provided script named `s2ui-create-challenge-questions`.  Please find more details in the <<scripts>> section
+
+All the answers should be encrypted and it will use the same encryption settings as the what is used to store the password
+for each user.  The service to handle this listener is automatically created and configured when
+you use the plugin.
+
+An Option to add  which can be configured as a list of items.  This will work regardless if e-mail validation is turned on.
+If the e-mail validation is turned on this step will occur before the e-mail is sent out.
+
+To make this work the domain object must have a link to your User Object. This is done by setting the `validationUserLookUpProperty` which if not set defaults to 'user'.
```
[the remark this answers is not in the document]
Comment (amondel2) [reply docs/src/docs/forgotPassword.adoc]: Both Fixed in commit

File: docs/src/docs/scripts/s2ui-create-challenge-questions.adoc [hunk 21]
```diff
@@ -0,0 +1,36 @@
...
+This script will create one controller which contains all the logic for handling the actions needed to provide a user interface to manage different users challenge questions and answers.
+
+This script will create two services.  The first service is to handle to be the implementation for the controller/domain object handling the data needed to set/get questions and answers.
+The second service is a listener service which ensures that anytime an answer is created or updated the value is encypted before it is stores in the database.
+
+This script will create three views.  Each of the three views (List/index, edit, and create) will support all the CRUD operations.
+
+This script will add lines into your application.groovy file in order to ensure the questions and answers are properly configured.
+
+This script will one line into i18n this will be the menu item in the nav bar for spring security.
```
[the remark this answers is not in the document]
Comment (amondel2) [reply docs/src/docs/scripts/s2ui-create-challenge-questions.adoc]: Fixed in next commit

File: docs/src/docs/userRegistration.adoc [hunk 22]
```diff
@@ -29,13 +32,31 @@ image::register_final.png[]
 
 The post-registration destination url is configurable in `grails-app/conf/application.groovy` using the `postRegisterUrl` attribute:
 
+In addition, each new user will be granted `ROLE_USER` after finalizing the registration. If you want to change the default role, add more, or grant no roles at all (for example if you want an admin to approve new users and explicitly enable new users) then you can customize that with the `defaultRoleNames` attribute (which is a List of Strings):
```
[the remark this answers is not in the document]
Comment (amondel2) [reply docs/src/docs/userRegistration.adoc]: I fixed this though I am not sure I wrote it!

File: docs/src/docs/userRegistration.adoc [hunk 23]
```diff
@@ -29,13 +32,31 @@ image::register_final.png[]
...
+----
+grails.plugin.springsecurity.ui.register.requireEmailValidation = false
+----
+
+*Please Note if you turn off e-mail validation in production it is strongly recommended to have additional user creation validation*
+
+
+=== Mail configuration
+
+By default the plugin uses the http://grails.org/plugin/mail[Mail] plugin to send emails, but only if it installed. This is configurable by registering your own `MailStrategy` implementation - see [the section on configuration|guide:customization] for more information. The plugin assumes that the Mail plugin and an SMTP server are already configured.
```
[the remark this answers is not in the document]
Comment (amondel2) [reply docs/src/docs/userRegistration.adoc]: Fixed in next Commit

[the remark this answers is not in the document]
Comment (amondel2) [reply docs/src/docs/userRegistration.adoc]: Fixed in next commit

File: docs/src/docs/forgotPassword.adoc [hunk 24]
```diff
@@ -38,8 +40,96 @@ The post-reset destination url is configurable in `grails-app/conf/application.g
 grails.plugin.springsecurity.ui.forgotPassword.postResetUrl = '/reset'
 ----
 
+=== Email Validation Configuration
 If you don't specify a value then the `defaultTargetUrl` value will be used, which is `'/'` by default.
```
[the remark this answers is not in the document]
Comment (amondel2) [reply docs/src/docs/forgotPassword.adoc]: Fixed in next commit

File: docs/src/docs/forgotPassword.adoc [hunk 25]
```diff
@@ -38,8 +40,96 @@ The post-reset destination url is configurable in `grails-app/conf/application.g
...
+[source,groovy]
+----
+grails.plugin.springsecurity.ui.forgotPassword.requireForgotPassEmailValidation = false
+----
+**If e-mail validation is turned off, it is recommended to use the forgotPasswordExtraValidation below.**
+
+
+=== Challenge Questions Configuration
+
+To use this option it is recommended to use the provided script named `s2ui-create-challenge-questions`.  Please find more details in the <<scripts>> section
```
[the remark this answers is not in the document]
Comment (amondel2) [reply docs/src/docs/forgotPassword.adoc]: Fixed in next commit

File: docs/src/docs/forgotPassword.adoc [hunk 26]
```diff
@@ -38,8 +40,96 @@ The post-reset destination url is configurable in `grails-app/conf/application.g
...
+
+All the answers should be encrypted and it will use the same encryption settings as the what is used to store the password
```
[the remark this answers is not in the document]
Comment (amondel2) [reply docs/src/docs/forgotPassword.adoc]: Fixed in next commit

**Comment by amondel2** [comment]
@ddelponte All bugs and Code Review items should be fixed.  Thank you very much for the review, please let me know if you find anything else that needs to be changed!

File: docs/src/docs/forgotPassword.adoc [hunk 27]
```diff
@@ -37,9 +39,99 @@ The post-reset destination url is configurable in `grails-app/conf/application.g
...
+|Header 1 |Header 2
+
+|labelDomain
+|This is the property in the domain object that will used to ask the question
+
+|labelMessage
+|If labelDomain is not present then this can be used as a message property to ask a question (ie if your questions are static).
+
+|prop
+|TThis is used as the property of the domain object to determine if the answer is correct.
```
Comment (ddelponte) [inline docs/src/docs/forgotPassword.adoc]: Apologies, but I found one more small typo... `TThis` should be `This`

**Comment by ddelponte** [comment]
Thanks again for all your hard work on this.  I have just one more bug to report.

1. I generated the challenge questions using the script as follows:

```
s2ui-create-challenge-questions com.mycompany.myapp ChallengeQuestion com.mycompany.myapp.User 2
```

2. the generated `ChallengeQuestionController` injects a `ChallengeQuestionService` with incorrect camel case

```
ChallengeQuestionService challengequestionService
```

should be

```
ChallengeQuestionService challengeQuestionService
```

image

**Comment by ddelponte** [comment]
Additional bugs:

1. The links generated in the `Profile/Challenge Question` drop down are incorrect
2. The create page of the `Profile/Challenge Questions` results in Internal server error

The video linked to below demonstrates how to reproduce the errors:

https://drive.google.com/file/d/1HCBoOqLOK5mZ2heqL_L77X80YYLjSzRp/view?usp=sharing

File: docs/src/docs/forgotPassword.adoc (hunk 27 above)
Comment (amondel2) [reply docs/src/docs/forgotPassword.adoc]: Fixed in next commit

**Comment by amondel2** [comment]
@ddelponte I believe all the bugs are fixed and they were related to camel casing.   The only thing that I could not reproduce is "The links generated in the Profile/Challenge Question drop down are incorrect" as even using the same script it produce no package name in the links and its just the controller.   I did add some code to fix the issue so it should not reproduce that issue.  It does mean getting the latest version of springSecurityUI.gsp

**Comment by ddelponte** [comment]
It's getting soooooo close!  Here's a bug I found with the latest code:

Excecuting the script 

```s2ui-create-challenge-questions com.mycompany.myapp ChallengeQuestion User 2```

Results in the wrong package being applied to the `User` domain class in the `ChallengeQuestion.groovy` domain class:

image

**Comment by amondel2** [comment]
What is the expected behavior if somebody does that?  I would assume it would add no package to the User class...but wanted to make sure it should not throw an error.

**Comment by ddelponte** [comment]
>I fixed this by assuming it would add no package (be the same as the challenge questions domain) to the User class. Let me know if you were expecting a different behavior.

I think it's okay to assume this, but would you mind updating the script documentation to recommend using the same package that contains the other spring security domain classes (i.e. the ones most often named `User`, `UserRole` and `Role`) when creating challenge questions?

**Comment by amondel2** [comment]
@ddelponte  that exists on line 25 of the script it self  and on line 15 of the documentation file scripts/s2ui-create-challenge-questions.adoc.  Are you expecting it somewhere else?

**Comment by ddelponte** [comment]
@amondel2 you're referring to this documentation, correct?

image

My thought was it may be beneficial to clarify in some way that the the script assumes the `User` domain class will reside in the same package as the generated challenge question.  If they reside in different packages, the developer will need to manually modify the generated challenge question domain class to properly reference the `User` domain class.

Maybe this is overkill.  I'm okay with whatever you decide to do.

**Comment by ddelponte** [comment]
@amondel2 no problem.  thanks!

**Comment by amondel2** [comment]
@ddelponte this is now in the script documentation file.  Hopefully there are not defects! :sweat_smile:

**Comment by ddelponte** [comment]
Things are looking great.  Here's one more nit-picky item (sorry :(  )

The Listener generated by the `s2ui-create-challenge-questions` script contains an unused import which may be deleted.

image

**Comment by amondel2** [comment]
Removed import from the Template and Listener created in the example program.

**Comment by amondel2** [comment]
@ddelponte I could not recreate your issues.  I followed the steps and got the expected results, there is even an Automated Geb test that does follow something similar to those steps.  I did create a video that demonstrated my testing I just did https://youtu.be/eBTEIbIrjoA, I did create a brand new project (that I am not checking in) based off of extended in the grails plugin.   If you can send me your project  where this is broken,  I can try and debug what happened.

**Comment by ddelponte** [comment]
Again, awesome work @amondel2 !  ✨ 

I've reviewed, tested and verified. Everything looks good to me.

I think this is ready for me to merge (I'll squash and merge it).

@sdelamo any objections?

**Comment by amondel2** [comment]
🍻  & 🍰
